// @applicaster/zapp-react-native-tvos-ui-components/Components/DefaultNavigation/styles.js
const BORDER_COLOR = "#FC461B";
const TEXT_COLOR = "#EFEFEF61";
const SELECTED_TEXT_COLOR = "#EFEFEF";
export const topBarContainer = {
  marginTop: 50,
  marginBottom: 20,
  alignItems: "center",
  justifyContent: "center",
};

export const topBarDataContainer = {
  marginHorizontal: 112,
  marginBottom: 10,
  flexDirection: "row",
  justifyContent: "space-between",
  alignItems: "center",
};

export const menuContainer = {
  flex: 1,
  height: 60,
};

export const navigationContainer = {
  height: 60,
  alignItems: "flex-start",
  justifyContent: "center",
  flexDirection: "row",
};

export const navigationCenterContainer = {
  ...navigationContainer,
  flex: 1,
};
export const navigationLeftContainer = {
  ...navigationContainer,
  alignItems: "center",
  width: 300,
};

export const navigationRightContainer = {
  ...navigationContainer,
  alignItems: "center",
  justifyContent: "flex-end",
  width: 300,
};

export const navItemImageOnlyStyle = {
  width: 300,
  height: 60,
};

export const navItemContent = {
  height: 60,
  marginHorizontal: 30,
  alignItems: "center",
  justifyContent: "center",
};

export const navItemContentFocused = {
  ...navItemContent,
  height: 66,
  borderColor: BORDER_COLOR,
  borderBottomWidth: 6,
};

export const navItemContentText = {
  fontSize: 24,
  fontWeight: "bold",
  fontStyle: "normal",
  color: TEXT_COLOR,
};

export const navItemContentTextSelected = {
  ...navItemContentText,
  color: SELECTED_TEXT_COLOR,
};

export const navItemIcon = {
  width: 50,
  height: 50,
  opacity: 0.3,
};

export const navItemIconFocused = {
  width: 50,
  height: 50,
};

export const navItemContentTextProps = {
  maxCharactersAllowed: 15,
  tailString: "...",
};
