// @applicaster/zapp-react-native-ui-components/Components/MasterCell/__tests__/testElements.js
export const elements = [
  {
    elements: [
      {
        elements: undefined,
        props: {
          uri: "http://urlOfTheImage.net/img.png",
        },
        style: {
          borderTopLeftRadius: 4,
          borderTopRightRadius: 4,
          height: 216,
          width: 384,
        },
        type: "Image",
      },
      {
        elements: [
          {
            elements: undefined,
            props: {
              label: "Test Entry",
              numberOfLines: 1,
            },
            style: {
              color: "#EFEFEF",
              fontSize: 26,
              fontStyle: "normal",
              fontWeight: "bold",
            },
            type: "Text",
          },
        ],
        props: {},
        style: {
          height: 30,
          marginHorizontal: 24,
          marginTop: 24,
          width: 336,
        },
        type: "View",
      },
      {
        elements: [
          {
            elements: undefined,
            props: {
              label: "Summary of the test entry",
              numberOfLines: 2,
            },
            style: {
              color: "#EFEFEF",
              fontSize: 24,
              fontStyle: "normal",
              fontWeight: "normal",
              opacity: 0.8,
            },
            type: "Text",
          },
        ],
        props: {},
        style: {
          height: 60,
          marginHorizontal: 24,
          marginTop: 10,
          width: 336,
        },
        type: "View",
      },
      {
        elements: [
          {
            elements: undefined,
            props: {
              uri: "http://urlOfTheImage.net/channelLogo.png",
            },
            style: {
              height: 42,
              width: 56,
            },
            type: "Image",
          },
        ],
        props: {},
        style: {
          height: 42,
          margin: 24,
          width: 56,
        },
        type: "View",
      },
    ],
    props: {},
    type: "View",
  },
];

export const singleElement = [
  {
    props: {},
    style: {
      backgroundColor: "#2F2F2F",
      borderRadius: 4,
      height: 436,
      marginRight: 20,
      width: 384,
    },
    type: "View",
  },
];
