// @applicaster/zapp-react-native-ui-components/Components/List/CellStyles/Rhine.js
const View = "View";
const Text = "Text";
const Image = "Image";

const styles = {
  container: {
    flex: 1,
    flexDirection: "row",
    height: 102,
  },
  mainImage: {
    width: 139,
    height: 78,
  },
  textContainer: {
    flex: 1,
    flexDirection: "column",
    justifyContent: "space-between",
    height: 78,
    paddingLeft: 10,
  },
  metaText: {
    fontSize: 12,
    height: 16,
    width: 179,
    marginBottom: 8,
    color: "#a0aab4",
  },
  title: {
    fontSize: 14,
    maxHeight: 54,
    fontWeight: "bold",
    color: "#414954",
  },
};

const viewTree = [
  {
    type: View,
    style: styles.container,
    elements: [
      {
        type: Image,
        style: styles.mainImage,
        data: [
          { key: ["media_group", 0, "media_item", 0, "src"], propName: "uri" },
        ],
      },
      {
        type: View,
        style: styles.textContainer,
        elements: [
          {
            type: Text,
            style: styles.metaText,
            data: [{ key: "summary", propName: "label" }],
            additionalProps: { numberOfLines: 1 },
          },
          {
            type: Text,
            style: styles.title,
            data: [{ key: "title", propName: "label" }],
            additionalProps: { numberOfLines: 3 },
          },
        ],
      },
    ],
  },
];

export const rhineCellStyle = {
  content_types: {
    default: {
      states: {
        default: viewTree,
      },
    },
  },
};
