// @applicaster/zapp-react-native-ui-components/Components/List/CellStyles/Rhine2.js
const View = "View";
const Text = "Text";
const Image = "Image";

const styles = {
  container: {
    flex: 1,
    paddingVertical: 12,
  },
  mainImage: {
    width: 351,
    height: 199,
    marginBottom: 16,
  },
  metaText: {
    fontSize: 12,
    height: 16,
    marginBottom: 8,
    color: "#a0aab4",
  },
  title: {
    fontSize: 14,
    maxHeight: 48,
    fontWeight: "bold",
    color: "#414954",
  },
};

const viewTree = [
  {
    type: View,
    style: styles.container,
    elements: [
      {
        type: Image,
        style: styles.mainImage,
        data: [
          { key: ["media_group", 0, "media_item", 0, "src"], propName: "uri" },
        ],
      },
      {
        type: Text,
        style: styles.metaText,
        data: [{ key: "summary", propName: "label" }],
        additionalProps: { numberOfLines: 1 },
      },
      {
        type: Text,
        style: styles.title,
        data: [{ key: "title", propName: "label" }],
        additionalProps: { numberOfLines: 2 },
      },
    ],
  },
];

export const rhineCellStyle2 = {
  content_types: {
    default: {
      states: {
        default: viewTree,
      },
    },
  },
};
