// @applicaster/zapp-react-native-ui-components/Components/MasterCell/DefaultComponents/Image.js
// @flow
import * as React from "react";
import { Image as RnImage } from "react-native";
import { basePath, fileExtension } from "../MappingFunctions/ImagePathMapper";
type Props = {
  style: any,
  uri: ?string,
  [string]: any,
  placeholderImage: ?string,
};

const placeholders = {
  "3.2": "placeholder_3.2",
};

function combinePath(filename) {
  return `${basePath}${filename}${fileExtension}`;
}

function getPlaceholder(placeholderImage, style) {
  const defaultSource =
    placeholderImage === undefined
      ? getPlaceholderForAspectRatio(style)
      : placeholderImage;

  if (placeholderImage && placeholderImage.uri) {
    return placeholderImage;
  } else if (defaultSource) {
    return { uri: defaultSource };
  }

  return null;
}

export function getPlaceholderForAspectRatio(style) {
  const { aspectRatio } = style;

  return combinePath(placeholders[aspectRatio] || "placeholder");
}

function Image({ style, uri, placeholderImage, ...otherProps }: Props) {
  const defaultSource = otherProps?.defaultSource;
  const placeholder = getPlaceholder(placeholderImage, style);

  return (
    <RnImage
      defaultSource={defaultSource}
      style={style}
      source={uri ? { uri } : placeholder}
      {...otherProps}
    />
  );
}

export default Image;
