// @applicaster/zapp-react-native-android-tv-ui-components/Components/ScreenPicker/ScreenSelector/styles.js
export const containerStyle = {
  flex: 1,
  width: 200,
};

export const masterCellContainerStyle = {
  flex: 1,
  height: 50,
};

const menuItem = styles => ({
  fontSize: 24,
  borderBottomWidth: 4,
  marginTop: 14,
  marginHorizontal: 14,
  paddingBottom: 10,
  color: styles.mainColor,
  borderColor: "transparent",
});

const menuItemSelected = styles => ({
  ...menuItem(styles),
  color: styles.highlightColor,
});

const menuItemFocused = styles => ({
  ...menuItem(styles),
  borderColor: styles.highlightColor,
});

const menuItemFocusedSelected = styles => ({
  ...menuItem(styles),
  color: styles.highlightColor,
  borderColor: styles.highlightColor,
});

export const screenSelectorItemStyle = { alignItems: "flex-start" };

export function menuItemStyle(selected, focused) {
  if (selected && focused) {
    return menuItemFocusedSelected;
  } else if (selected) {
    return menuItemSelected;
  } else if (focused) {
    return menuItemFocused;
  } else {
    return menuItem;
  }
}
