// @applicaster/zapp-react-native-ui-components/Components/Focusable/index.android.js
// @flow
import * as React from "react";
import { useFocusManager } from "@applicaster/zapp-react-native-utils/focusManager";
import { Touchable } from "./Touchable";

type Props = {
  children: React.Node,
  id: String,
  nextFocusUp: React.Ref,
  nextFocusRight: React.Ref,
  nextFocusLeft: React.Ref,
  nextFocusDown: React.Ref,
  hasInitialFocus: boolean,
  onFocus: () => void,
  onBlur: () => void,
  onPress: () => void,
};

function FocusableComponent(props: Props, ref) {
  const {
    children,
    id,
    nextFocusDown,
    nextFocusUp,
    nextFocusLeft,
    nextFocusRight,
    hasInitialFocus,
    onFocus,
    onBlur,
    onPress,
  } = props;

  const focusManager = useFocusManager();

  const extractCurrent = (ref) => {
    return ref?.current || ref;
  };

  React.useEffect(() => {
    hasInitialFocus && focusManager.setFocus(ref);
  }, [hasInitialFocus]);

  const isFocused = focusManager.isFocused(id);

  return (
    <Touchable
      ref={ref}
      {...{
        nextFocusDown: extractCurrent(nextFocusDown),
        nextFocusUp: extractCurrent(nextFocusUp),
        nextFocusLeft: extractCurrent(nextFocusLeft),
        nextFocusRight: extractCurrent(nextFocusRight),
        id,
        onFocus,
        onBlur,
        onPress,
      }}
    >
      {children(
        isFocused,
        {
          nextFocusDown,
          nextFocusUp,
          nextFocusLeft,
          nextFocusRight,
        },
        ref
      )}
    </Touchable>
  );
}

export const Focusable = React.forwardRef(FocusableComponent);
