// @applicaster/zapp-pipes-dev-kit/lib/config/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var env = process.env.NODE_ENV || "development";

function retrieveEnv() {
  switch (env) {
    case "development":
      return require("./development/index").default;
    case "test":
      return require("./test/index").default;
    case "production":
      return require("./production/index").default;
    default:
      return {};
  }
}

var envConfig = retrieveEnv();
var baseConfig = {
  errorMessages: {
    unknownProvider: function unknownProvider(provider) {
      return "no " + provider + " provider found";
    },
    noHandler: function noHandler(provider) {
      return "provider " + provider + " has no handler";
    },
    noManifest: function noManifest(provider) {
      return "provider " + provider + " has no manifest";
    },
    unknownCommand: function unknownCommand(provider, command) {
      return "provider " + provider + " doesn't know how to handle command " + command;
    }
  }
};

exports.default = _extends({}, baseConfig, envConfig);
