// @applicaster/zapp-react-dom-app/App/Layout/Layout.js
// @flow

import * as React from "react";
import { View } from "react-native";

import { Background } from "@applicaster/zapp-react-dom-ui-components/Components/Background";
import { NavWrapper } from "@applicaster/zapp-react-dom-ui-components/Components/NavWrapper";

import { getLocalizations } from "./layoutHelpers";

type Props = {
  children: React.ComponentType<any>,
  plugins: [{ module: any, type: string, name: string }],
  rivers: { [string]: any },
  localizations: { [string]: string | {} },
  appState: { appReady: boolean },
  navigator: {
    activeRiver: { id: string, [string]: any },
    isNavBarVisible: boolean,
  },
};

const displayStateStyles = {
  overflow: "hidden",
  flex: 1,
  flexDirection: "column",
  justifyContent: "flex-start",
  minHeight: "100vh",
};

export function LayoutComponent(props: Props) {
  const { children, navigator } = props;
  const { appState } = props;
  const { appReady = false } = appState;
  const { activeRiver, isNavBarVisible } = navigator;

  const localizations = getLocalizations(props);

  return (
    appReady && (
      <View>
        <Background localizations={localizations} />
        <View style={displayStateStyles}>
          {isNavBarVisible && <NavWrapper activeRiver={activeRiver} />}
          {children}
        </View>
      </View>
    )
  );
}
