// @applicaster/zapp-react-native-android-tv-ui-components/Components/HorizontalList/styles.js
const styles = {
  MAIN_MENU: {
    listStyle: "none",
    margin: "0",
    paddingLeft: "0px",
    paddingBottom: "80px",
    float: "left",
    width: "1000px"
  },
  CHANNEL_MENU: {
    listStyle: "none",
    margin: "0",
    paddingLeft: "400px",
    width: "100%",
    float: "left"
  },
  CATEGORIES_MENU: {
    listStyle: "none",
    margin: "0",
    padding: "0px 30px 0 60px",
    width: "398px",

    float: "left"
  },
  SETTINGS_MENU: {
    listStyle: "none",
    margin: "0",
    padding: "0px 30px 0 100px",
    width: "398px",
    height: "1000px",

    float: "left"
  },
  FAMILY_DANUBE_LIST_1: {
    fontFamily: "Montserrat",
    listStyle: "none",
    margin: "0",
    paddingLeft: "117px",
    paddingBottom: "30px",
    position: "relative",
    float: "left",
    whiteSpace: "normal",
    width: "200000px",
    height: ""
  },
  FAMILY_DANUBE_LIST_2: {
    fontFamily: "Montserrat",
    listStyle: "none",
    margin: "0",
    paddingLeft: "117px",
    paddingBottom: "30px",
    position: "relative",
    float: "left",
    whiteSpace: "normal",
    width: "200000px"
  },
  FAMILY_DANUBE_LIST_3: {
    fontFamily: "Montserrat",
    listStyle: "none",
    margin: "0",
    paddingLeft: "117px",
    paddingBottom: "30px",
    position: "relative",
    float: "left",
    whiteSpace: "normal",
    width: "200000px"
  },
  FAMILY_DANUBE_HORIZONTAL_LIST_1: {
    fontFamily: "Montserrat",
    listStyle: "none",
    margin: "0",
    paddingLeft: "117px",
    paddingBottom: "30px",
    position: "relative",
    float: "left",
    whiteSpace: "normal",
    width: "200000px",
    height: "428px"
  },
  FAMILY_DANUBE_HORIZONTAL_LIST_2: {
    fontFamily: "Montserrat",
    listStyle: "none",
    margin: "0",
    paddingLeft: "117px",
    paddingBottom: "30px",
    position: "relative",
    float: "left",
    whiteSpace: "normal",
    width: "200000px",
    height: "428px"
  },
  FAMILY_DANUBE_HORIZONTAL_LIST_3: {
    fontFamily: "Montserrat",
    listStyle: "none",
    margin: "0",
    paddingLeft: "117px",
    paddingBottom: "30px",
    position: "relative",
    float: "left",
    whiteSpace: "normal",
    width: "200000px",
    height: "200px"
  },
  FEATURED_RAIL_POPULAR: {
    listStyle: "none",
    margin: "0",
    paddingLeft: "117px",
    paddingBottom: "30px",
    position: "relative",
    float: "left",
    whiteSpace: "normal",
    width: "200000px"
  },
  HEADER_1: {
    width: "1696px",
    height: "60px",
    fontFamily: "Montserrat",
    fontSize: "34px",
    fontWeight: "bold",
    fontStyle: "normal",
    fontStretch: "normal",
    lineHeight: "1.76",
    letterSpacing: "normal",
    color: {
      defaultColor: "#efefef",
      key: "main_text"
    }
  }
};

export default styles;
