// @applicaster/zapp-react-native-ui-components/Components/PlayerController/ErrorDisplay/ErrorDisplay.js
// @flow
import * as React from "react";
import * as R from "ramda";
import { View, Text } from "react-native";

import { getLocalizations } from "@applicaster/zapp-react-native-tvos-app/Utils/localizationsHelper";
import { getAppStylesColor } from "@applicaster/zapp-react-native-utils/stylesUtils";

type Props = {
  localizations: {},
  styles: {},
};

const defaultAppStyles = {
  loading_error_label: {
    color: "#ffefefef",
  },
};

const textStyles = (appStyles = defaultAppStyles) => ({
  color: getAppStylesColor("loading_error_label", appStyles),
  fontSize: 36,
});

const errorStyles = {
  flex: 1,
  width: "100%",
  alignItems: "center",
  justifyContent: "flex-start",
  paddingTop: 200,
};

export function ErrorDisplayComponent({
  localizations,
  styles,
}: Props) {
  const { stream_error_message = "Cannot play stream" } = getLocalizations({
    localizations,
  });

  const appStyles = R.prop("universal", styles);
  return (
    <View style={errorStyles}>
      <Text style={textStyles(appStyles)}>{stream_error_message}</Text>
    </View>
  );
}
