// @applicaster/zapplicaster-cli/src/commands/prepareWorkspace/index.js
/**
 * module prepareWorkspace
 * @module @applicaster/zapplicaster-cli/src/commands/prepareWorkspace
 * This module contains the code which runs the prepare command of the
 * zapplicaster-cli
 */

const { taskRunner } = require("../../taskRunner");

const { prerequisitesChecker } = require("./prerequisitesChecker");
const { configurator } = require("./configurator");
const { workspaceCreator } = require("./workspaceCreator");
const { configFilesRetriever } = require("./configFilesRetriever");
const {
  runtimeConfigurationUrlsCreator,
} = require("./runtimeConfigurationUrlsCreator");
const { appBootstrapper } = require("./appBootstrapper");

const prepareWorkspaceTask = {
  name: "prepare workspace",
  startMessage:
    "We're preparing your app - you'll be happily coding in no time !",
  prerequisitesChecker,
  configurator,
  steps: [
    {
      start: "Creating workspace",
      run: workspaceCreator,
      error: "Could not create workspace",
      completion: "Workspace created !",
    },
    {
      start: "Retrieving configuration files",
      run: configFilesRetriever,
      error: "Could not retrieve configuration files",
      completion: "Configuration files retrieved !",
    },
    {
      start: "Creating runtime URLs configuration file",
      run: runtimeConfigurationUrlsCreator,
      error: "Could not create runtime_configuration_urls.json",
      completion: "Runtime configuration URLs file created",
    },
    {
      start: "Bootstraping app project",
      run: appBootstrapper,
      error: "Could not bootstrap the app",
      completion: "Your app is ready !",
    },
  ],
};

module.exports = { prepareWorkspace: taskRunner(prepareWorkspaceTask) };
