// @applicaster/zapplicaster-cli/src/commands/migrateZappPlugin/index.js
/**
 * module migrateZappPlugin
 * @module @applicaster/zapplicaster-cli/src/commands/migrateZappPlugin
 * This module automatically update a legacy Zapp RN plugin to work with
 * QuickBrick
 */

const { taskRunner } = require("../../taskRunner");
const { configurator } = require("./configurator");
const { updatePackageJson } = require("./updatePackageJson");
const { createPluginEntryPoint } = require("./createPluginEntryPoint");
const { updateZappifest } = require("./updateZappifest");

const migrateZappPluginTask = {
  name: "migrate Zapp Plugin",
  startMessage: "We're updating your plugin to be compatible with QuickBrick",
  configurator,
  steps: [
    {
      start: "Updating package.json",
      run: updatePackageJson,
      error: "Could not update your package.json",
      completion: "Done !",
    },
    {
      start: "Creating plugin entry point",
      run: createPluginEntryPoint,
      error: "Could not create the plugin entry point",
      completion: "Done !",
    },
    {
      start: "Updating Zappifest",
      run: updateZappifest,
      error:
        "Could not update Zappifest ! please update it manually by refering to https://developer-zapp.applicaster.com/quick-brick/migration-guide.html", // eslint-disable-line max-len
      completion:
        "Your plugin is ready ! Simply publish it to npm, and it will be ready to use in any QuickBrick app",
    },
  ],
};

module.exports = { migrateZappPlugin: taskRunner(migrateZappPluginTask) };
