// @applicaster/zapp-react-native-utils/analyticsUtils/events.js
export const ANALYTICS_CORE_EVENTS = {
  TAP_CELL: "Tap Cell",
  PLAY_VOD_ITEM: "Play VOD Item",
  VOD_ITEM_PLAY_WAS_TRIGGERED: "VOD Item: Play Was Triggered",
  CHANNEL_ITEM_PLAY_WAS_TRIGGERED: "Channel Item: Play Was Triggered",
  PROGRAM_ITEM_PLAY_WAS_TRIGGERED: "Program Item: Play Was Triggered",
  TAP_NAVBAR_ICON: "Tap Navbar Icon", // Not supported
  TAP_NAVBAR_BACK_BUTTON: "Tap Navbar Back Button", // Should we use this key?
  HOME_SCREEN_VIEWED: "Home Screen: Viewed",
  LAUNCH_APP: "Launch App"
};

export const ANALYTICS_LAUNCH_EVENTS = {
  VERSION_NAME: "Version Name",
  BUNDLE_VERSION: "Bundle Version",
  RIVERS_CONFIGURATON_ID: "Rivers Configuration ID",
  UUID: "uuid",
  MORPHEOUS_VERSION: "Morpheous Version"
};

export const ANALYTICS_ENTRY_EVENTS = {
  ITEM_TYPE: "Item Type",
  ITEM_TITLE: "Item Name",
  ITEM_ID: "Item ID",
  ITEM_LINK: "Item Link",
  ITEM_TARGET: "Item Target",
  ITEM_INDEX: "Item Number",
  ITEM_INDEX_IS_FREE: "Free or Paid Video",
  ITEM_DURATION: "Item Duration",
  ITEM_PLAY_COMPLETED: "Completed",
  ITEM_CUSTOM_PROPERTY: "Custom Property"
};

export const ANALYTICS_COMPONENT_EVENTS = {
  COMPONENT_ID: "Component ID",
  COMPONENT_TYPE: "Component Type",
  CELL_STYLE: "Cell Style",
  CELL_FAMILY: "Cell Family",
  HAS_HEADERS: "Has Header",
  HEADER_TITLE: "Header Name",
  CONNECTED_SCREEN: "Connected Screen",
  COMPONENT_TARGET: "Component Target",
  COMPONENT_SOURCE: "Component Source",
  COMPONENT_SOURCE_TYPE: "Component Source Type",
  ITEM_LIMIT: "Item Limit"
};
