// @applicaster/zapp-react-dom-ui-components/Components/BackToVideo/BackToVideo.js
// @flow
import * as React from "react";
import * as R from "ramda";
import { View, Text } from "react-native";

import { getLocalizations } from "@applicaster/zapp-react-native-utils/appUtils/localizationsHelper";

import { FocusableGroup } from "@applicaster/zapp-react-native-ui-components/Components/FocusableGroup";
import { Focusable } from "@applicaster/zapp-react-native-ui-components/Components/Focusable";
import { DISPLAY_STATES } from "../../Contexts/DisplayStateContext";

import {
  verticalOffset,
  defaultResumeVideoString,
  defaultNowPlayingString,
} from "./consts";

import {
  containerStyles,
  nowPlayingStyles,
  titleStyles,
  buttonStyles,
  textButtonStyles,
} from "./styles";

type Props = {
  playerContent: {
    title: string,
  },
  setDisplayState: string => void,
  riverOffsetUpdater: (string, number) => void,
  styles: {
     styles: {
      tv: { [string]: { color: string } },
    },
  },
  localizations: { [string]: { [string]: string } },
};

export function BackToVideoComponent({
  playerContent,
  setDisplayState,
  riverOffsetUpdater,
  styles,
  localizations,
}: Props) {
  if (!playerContent) {
    return null;
  }

  const appStyles = R.prop("tv", styles);

  const {
    resume_video_text = defaultResumeVideoString,
    now_playing_text = defaultNowPlayingString,
  } = getLocalizations({ localizations });

  const { title } = playerContent;
  const id = "backToVideo";
  
  return (
    <FocusableGroup id={id}>
      <View style={containerStyles}>
        <Text style={nowPlayingStyles(appStyles)}>
          {R.toUpper(now_playing_text)}
        </Text>
        <Text style={titleStyles(appStyles)}>{title}</Text>
        <Focusable
          groupId={id}
          id={id}
          onFocus={focusable =>
            riverOffsetUpdater(
              "vertical",
              verticalOffset - focusable.getRect().top
            )
          }
          onPress={() => setDisplayState(DISPLAY_STATES.PLAYER)}
        >
          {focused => (
            <View style={buttonStyles(focused, appStyles)}>
              <Text style={textButtonStyles(focused, appStyles)}>
                {R.toUpper(resume_video_text)}
              </Text>
            </View>
          )}
        </Focusable>
      </View>
    </FocusableGroup>
  );
}
