// @applicaster/zapp-react-dom-ui-components/Components/BackToVideo/styles.js
import { getAppStylesColor } from "@applicaster/zapp-react-native-utils/stylesUtils";

const defaultAppStyles = {
  active: { color: "#fffc461b" },
  main_text: { color: "#ffefefef" },
};

export const containerStyles = {
  flex: 1,
  height: 420,
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "center",
  paddingBottom: 48,
  paddingTop: 96,
};

export const nowPlayingStyles = (appStyles = defaultAppStyles) => ({
  fontSize: 32,
  fontWeight: "bold",
  color: getAppStylesColor("active", appStyles),
  textAlign: "center",
});

export const titleStyles = (appStyles = defaultAppStyles) => ({
  height: 90,
  fontSize: 70,
  marginTop: 10,
  fontWeight: "bold",
  color: getAppStylesColor("main_text", appStyles),
  textAlign: "center",
});

export const buttonStyles = (focused, appStyles = defaultAppStyles) => ({
  width: 242,
  height: 70,
  marginTop: 60,
  borderRadius: 4,
  borderWidth: 2,
  borderColor: focused
    ? getAppStylesColor("active", appStyles)
    : getAppStylesColor("main_text", appStyles),
  opacity: focused ? 1 : 0.5,
  backgroundColor: focused
    ? getAppStylesColor("active", appStyles)
    : "transparent",
  alignItems: "center",
  justifyContent: "center",
});

export const textButtonStyles = (focused, appStyles = defaultAppStyles) => ({
  fontSize: 24,
  fontWeight: "bold",
  opacity: focused ? 1 : 0.5,
  color: getAppStylesColor("main_text", appStyles),
});
