// @applicaster/zapp-react-dom-ui-components/Components/Cell/CellStyles/ScreenSelector/index.js
const viewTree = state => [
  {
    type: "View",
    style: {
      flex: 1,
      // aspectRatio: 1.0487804878,
      paddingTop: 14,
      paddingBottom: 14,
      paddingRight: 30,
    },
    elements: [
      {
        type: "View",

        elements: [
          {
            type: "Text",

            style: {
              // paddingHorizontal: "15%",
              paddingTop: "2%",
              paddingBottom: "2%",
              fontSize: 26,
              lineHeight: 32,
              color:
                state === "focused" || state === "selected"
                  ? "rgba(239,239,239,1.0)"
                  : "rgba(239,239,239,0.61)",
              fontWeight: "700",
            },
            data: [
              {
                func: "path",
                args: ["title"],
                propName: "label",
              },
            ],
            additionalProps: {
              numberOfLines: 2,
            },
          },
          {
            type: "View",
            style: {
              height: 10,
              alignItems: "center",
              justifyContent: "center",
              borderColor: state === "focused" ? "#FC461B" : "transparent",
              borderBottomWidth: 6,
            },
          },
        ],
      },
    ],
  },
];

export const screenSelector = {
  content_types: {
    default: {
      states: {
        default: viewTree("default"),
        focused: viewTree("focused"),
        selected: viewTree("selected"),
      },
    },
  },
};
