// @applicaster/zapp-react-native-android-tv-ui-components/Components/Collection/Collection.js
// @flow
import * as React from "react";

import { Cell } from "../Cell";
import { ContainerWrapperView } from "../ContainerWrapperView";

import { FlatList } from "react-native";

type Props = {
  items: [],
  componentId: string,
  component: { [string]: any },
  CellRenderer: React.componentType<{ item: Object, state: string }>,
  styles: {},
  shouldScrollVertically: ?([any]) => boolean,
  shouldScrollHorizontally: ?([any]) => boolean,
  navigator: {},
  verticalAnimationOffset: ?number,
  hasHeader: Boolean,
  numColumns: ?number,
  horizontal: ?boolean,
};

function keyExtractor(item) {
  return item.toString();
}

export function CollectionComponent({
  items,
  componentId,
  CellRenderer,
  styles,
  shouldScrollVertically,
  shouldScrollHorizontally,
  navigator,
  component,
  verticalAnimationOffset = 0,
  hasHeader,
  numColumns,
  horizontal,
}: Props) {
  return (
    <ContainerWrapperView style={styles}>
      {() => (
        <FlatList
          horizontal={horizontal}
          numColumns={numColumns}
          removeClippedSubviews={true}
          data={items}
          keyExtractor={(item, index) => keyExtractor(index)}
          renderItem={({ item, index }) => {
            return (
              <Cell
                componentId={componentId}
                CellRenderer={CellRenderer}
                item={item}
                key={index}
                component={component}
                shouldScrollHorizontally={shouldScrollHorizontally}
                shouldScrollVertically={shouldScrollVertically}
                hasPreferredFocus={index === 0}
                navigator={navigator}
                hasHeader={hasHeader}
                verticalAnimationOffset={verticalAnimationOffset}
              />
            );
          }}
        />
      )}
    </ContainerWrapperView>
  );
}
