// @applicaster/zapp-react-dom-ui-components/Components/Background/Background.js
// @flow
/* eslint-disable no-console */
import * as React from "react";
import * as R from "ramda";

import { BackgroundImage } from "./BackgroundImage";

type Props = {
  Player: React.ComponentType,
  playerContent: ?{},
  styles: {
    tv: {
      background: string,
    },
    
  },
  remoteConfigurations: {
    styles: {
      tv: {
        background: string,
      },
    },
  },
  displayState: string,
};

export function BackgroundComponent({ remoteConfigurations, styles }: Props) {
  const localAssets = R.either(
    R.path(["sdkAssets", "bgImage"]),
    R.path(["sdkAssets", "app_background_image_1080p"])
  );
  const remoteAssets = R.either(
    R.path(["assets", "universal", "bgImage"]),
    R.path(["assets", "universal", "app_background_image_1080p"])
  );

  const background_color = R.path(
    [ "tv", "background_color", "color" ],
    styles
  );

  const backgroundImageUrl = R.either(
    localAssets,
    remoteAssets
  )(remoteConfigurations);

  return (
    <BackgroundImage
      imageUrl={backgroundImageUrl}
      backgroundColor={background_color}
    />
  );
}
