// @applicaster/zapp-react-native-tvos-app/Layout/Layout.js
// @flow
import * as React from "react";
import { BackgroundImage } from "@applicaster/zapp-react-native-tvos-ui-components/Components/BackgroundImage";
import { RootViewWrapper } from "../RootViewWrapper";
import { isScreenPlayable } from "@applicaster/zapp-react-native-utils/navigationUtils/itemTypes";

type Props = {
  activeRiver: any,
  navigator: any,
  children: React.ComponentType<any>,
  plugins: [{ module: any, type: string, name: string }],
  rivers: { [string]: any },
  appState: { appReady: boolean },
};

export function LayoutComponent(props: Props) {
  const {
    appState: { appReady },
    navigator,
    rivers,
    children,
  } = props;

  const { activeRiver, pathname } = navigator;

  const isPlayerPresented = isScreenPlayable(pathname);

  return (
    <BackgroundImage>
      {appReady ? (
        <RootViewWrapper
          location={{ pathname }}
          activeRiver={activeRiver}
          navigator={navigator}
          rivers={rivers}
          isPlayerPresented={isPlayerPresented}
        >
          {children}
        </RootViewWrapper>
      ) : null}
    </BackgroundImage>
  );
}
