// @applicaster/zapp-react-native-utils/stylesUtils/Keys/index.js
import { platformSelect } from "../../reactUtils";

export const STYLES = {
  BACKGROUND_COLOR: platformSelect({
    samsung_tv: "background",
    lg_tv: "background",
    tvos: "background_color",
    ios: "background_color",
    android_tv: "tv_app_background_color",
    android: "tv_app_background_color",
    default: "background_color",
  }),
  ACTIVE: platformSelect({
    samsung_tv: "active",
    lg_tv: "active",
    tvos: "active",
    ios: "active",
    android: "tv_brand_color",
    android_tv: "tv_brand_color",
    default: "active",
  }),
  MAIN_TEXT: platformSelect({
    samsung_tv: "main_text",
    lg_tv: "main_text",
    tvos: "main_text",
    ios: "main_text",
    android: "header_primary_text",
    android_tv: "header_primary_text",
    default: "main_text",
  }),
  FOCUSED_TEXT_COLOR: platformSelect({
    samsung_tv: "focused_text_color",
    lg_tv: "focused_text_color",
    tvos: "focused_text_color",
    ios: "focused_text_color",
    android: "tv_brand_color",
    android_tv: "tv_brand_color",
    default: "focused_text_color",
  }),
  HEADER_TEXT_COLOR: platformSelect({
    samsung_tv: "header_text_color",
    lg_tv: "header_text_color",
    tvos: "header_text_color",
    ios: "header_text_color",
    android: "TvHeaderTitle",
    android_tv: "TvHeaderTitle",
    default: "header_text_color",
  }),
};

export const ASSETS = {
  APP_BACKGROUND_IMAGE: platformSelect({
    samsung_tv: "app_background_image_1080p",
    lg_tv: "bgImage",
    tvos: "app_background_image",
    ios: "app_background_image",
    android_tv: "tv_app_background",
    android: "",
    default: "app_background_image",
  }),
  PLACEHOLDER: platformSelect({
    samsung_tv: "placeholder",
    lg_tv: "placeholder",
    tvos: "placeholder",
    ios: "placeholder",
    android: "placeholder",
    android_tv: "placeholder",
    default: "placeholder",
  }),
  PLACEHOLDER_3_2: platformSelect({
    samsung_tv: "placeholder_3.2",
    lg_tv: "placeholder_3.2",
    tvos: "placeholder_3.2",
    ios: "placeholder_3.2",
    android: "placeholder_3_2",
    android_tv: "placeholder_3_2",
    default: "placeholder_3.2",
  }),
  SPLASH_SCREEN: platformSelect({
    samsung_tv: "splash",
    lg_tv: "splashBackground",
    tvos: "launch",
    ios: "launch",
    android: "",
    default: "splash",
  }),
};
