// @applicaster/zapp-react-native-ui-components/Components/Header/index.js
// @flow
import * as React from "react";
import * as R from "ramda";
import { View, Text } from "react-native";
import { connectToStore } from "@applicaster/zapp-react-native-redux";
import { platformSelect } from "@applicaster/zapp-react-native-utils/reactUtils";
import { fixColorHexCode } from "@applicaster/zapp-react-native-utils/stylesUtils";

type Props = {
  title: string,
  styles: {},
}

const HeaderComponent = ({
  styles,
  title,
}: Props) => {
  if (R.isNil(title) || R.isEmpty(title)) {
    return null;
  }

  const underline_col = fixColorHexCode(R.path(["smartphone", "app_background_color", "color"], styles));
  const underlineColor = underline_col || "rgba(104,104,104,1.0)";

  const headerContainer = {
    constainter: {
      marginTop: 12,
      padding: 12,
    },
    underline: {
      width: 100,
      height: 5,
      backgroundColor: underlineColor,
    },
  };

  const headerFontStyles = {
    fontSize: platformSelect({
      ios: Number(R.path(["iphone", "header_font_style", "font_size"], styles)),
      android: Number(R.path(["smartphone", "header_font_style", "font_size"], styles)),
    }),
    color: platformSelect({ios: fixColorHexCode(R.path(["iphone", "header_font_style", "color"], styles)),
      android: fixColorHexCode(R.path(["smartphone", "header_font_style", "color"], styles)),
    }),
    fontFamily: platformSelect({
      ios: R.path(["iphone", "header_font_style", "font"], styles),
      android: R.path(["smartphone", "header_font_style", "font"], styles),
    }),
  };
  
  return (
    <View style={headerContainer.container}>
      <Text style={headerFontStyles}>{title}</Text>
      <View style={headerContainer.underline} />
    </View>
  );
};

export const Header = connectToStore(
  R.pick(["styles", "remoteConfigurations"])
)(HeaderComponent);
