// @applicaster/zapp-react-native-android-tv-ui-components/Components/Cell/styles.js
const BACKGROUND_COLOR = "#2F2F2F";
const SELECTED_BACKGROUND_COLOR = "#FC461B";
const FOCUSED_BACKGROUND_COLOR = "#FC461B";
const SUBTITLE_COLOR1 = "#EFEFEF";
const TITLE_COLOR1 = "#EFEFEF";

export const cellStyle1 = {
  container: {
    width: 384,
    height: 372,
    borderRadius: 4,
    backgroundColor: BACKGROUND_COLOR,
    marginRight: 48,
    marginBottom: 48,
  },
  selectedContainer: {
    width: 384,
    height: 372,
    borderRadius: 4,
    backgroundColor: SELECTED_BACKGROUND_COLOR,
    marginRight: 48,
    marginBottom: 48,
  },
  focusedContainer: {
    width: 384,
    height: 372,
    borderRadius: 4,
    backgroundColor: FOCUSED_BACKGROUND_COLOR,
    marginRight: 48,
    marginBottom: 48,
  },
  imageContainer: {
    width: 384,
    height: 216,
    borderTopRightRadius: 4,
    borderTopLeftRadius: 4,
  },
  image: {
    width: 384,
    height: 216,
    borderTopRightRadius: 4,
    borderTopLeftRadius: 4,
  },
  titleContainer: {
    width: 336,
    height: 30,
    marginHorizontal: 24,
    marginTop: 24,
  },
  subTitleContainer: {
    width: 336,
    height: "auto",
    marginHorizontal: 24,
    marginTop: 10,
  },
  subTitle: {
    fontSize: 24,
    fontWeight: "normal",
    opacity: 0.8,
    color: SUBTITLE_COLOR1,
    fontStyle: "normal",
    fontFamily: "Montserrat",
  },
  title: {
    fontSize: 26,
    color: TITLE_COLOR1,
    fontWeight: "bold",
    fontStyle: "normal",
    fontFamily: "Montserrat",
  },
};

export const cellStyle2 = {
  container: {
    width: 384,
    height: 372,
    borderRadius: 4,
    backgroundColor: BACKGROUND_COLOR,
    marginRight: 48,
    marginBottom: 48,
  },
  selectedContainer: {
    width: 384,
    height: 372,
    borderRadius: 4,
    backgroundColor: SELECTED_BACKGROUND_COLOR,
    marginRight: 48,
    marginBottom: 48,
  },
  focusedContainer: {
    width: 384,
    height: 372,
    borderRadius: 4,
    backgroundColor: FOCUSED_BACKGROUND_COLOR,
    marginRight: 48,
    marginBottom: 48,
  },
  imageContainer: {
    width: 384,
    height: 216,
    borderTopRightRadius: 4,
    borderTopLeftRadius: 4,
  },
  image: {
    width: 384,
    height: 216,
    borderTopRightRadius: 4,
    borderTopLeftRadius: 4,
  },
  titleContainer: {
    width: 336,
    height: 30,
    marginHorizontal: 24,
    marginTop: 24,
  },
  subTitleContainer: {
    width: 336,
    height: "auto",
    marginHorizontal: 24,
    marginTop: 10,
  },
  subTitle: {
    fontSize: 24,
    fontWeight: "normal",
    opacity: 0.8,
    color: SUBTITLE_COLOR1,
    fontStyle: "normal",
    fontFamily: "Montserrat",
  },
  title: {
    fontSize: 26,
    color: TITLE_COLOR1,
    fontWeight: "bold",
    fontStyle: "normal",
    fontFamily: "Montserrat",
  },
};
