// @applicaster/zapp-react-dom-ui-components/Components/Grid/Grid.js
// @flow
import * as React from "react";
import * as R from "ramda";

import { Collection } from "../Collection";
import { FocusableGroup } from "@applicaster/zapp-react-native-ui-components/Components/FocusableGroup";
const gridStyle = {
  flex: 1,
  flexDirection: "row",
  display: "flex",
  marginTop: 12,
  marginLeft: 114,
  flexWrap: "wrap",
};

type Props = {
  component: {
    id: string,
    styles: { header: {} },
  },
  componentIndex: number,
  CellRenderer: React.componentType<{ item: Object, state: string }>,
  zappPipesData: {
    loading: ?boolean,
    data: {
      title: string,
      entry: [{}],
    },
  },
  groupId: String,
  isScreenWrappedInContainer: boolean,
};

export function GridComponent({
  component,
  componentIndex,
  CellRenderer,
  zappPipesData,
  groupId,
  isScreenWrappedInContainer,
}: Props) {
  const {
    data: { title, entry },
  } = zappPipesData;

  if (entry.length === 0) {
    return null;
  }

  const {
    id,
    styles: { header },
  } = component;

  const headerData = R.merge(header, { title });

  return (
    <FocusableGroup
      id={id}
      groupId={groupId}
      preferredFocus={componentIndex === 0}
      shouldUsePreferredFocus
    >
      <Collection
        componentId={id}
        component={component}
        CellRenderer={CellRenderer}
        items={entry}
        header={headerData}
        styles={gridStyle}
        shouldScrollVertically={R.T}
        isScreenWrappedInContainer={isScreenWrappedInContainer}
      />
    </FocusableGroup>
  );
}
