// @applicaster/zapp-react-native-ui-components/Components/Favorites/FavoritesComponent.js
// @flow
import * as React from "react";
import { View, Text, StyleSheet } from "react-native";

import { useTheme } from "@applicaster/zapp-react-native-utils/theme";
import { withNavigator } from "../../Decorators/Navigator";

import { ComponentResolver } from "../ComponentResolver";
import { assignDataSource } from "./assignDataSource";

const eightyPercentBlack = "rgba(0,0,0, 0.8)";

const styles = StyleSheet.create({
  container: (theme) => ({
    flex: 1,
    backgroundColor: theme.app_background_color,
    alignItems: "center",
    justifyContent: "center",
  }),
  text: {
    color: eightyPercentBlack,
  },
});

type Props = {
  component: {
    id: string,
  },
  feed: {
    entry: [],
  },
};

export function FavoritesComponent({ component, feed }: Props) {
  const theme = useTheme();

  if (feed?.entry?.length > 0) {
    return (
      <ComponentResolver
        key={feed?.entry.length}
        component={component}
        decorators={[withNavigator, assignDataSource(feed)]}
      >
        {(Component, CellRenderer) => {
          return (
            <Component
              key={component.id}
              isScreenWrappedInContainer={false}
              component={component}
              componentIndex={0}
              CellRenderer={CellRenderer}
              onLoadFinished={() => {}}
            />
          );
        }}
      </ComponentResolver>
    );
  }

  return (
    <View style={styles.container(theme)}>
      <Text style={styles.text}>You don't have any favourites</Text>
    </View>
  );
}
