// @applicaster/zapplicaster-cli/src/commands/runApp/runApp.js
/* eslint-disable max-len */
/* eslint-disable no-template-curly-in-string */
const fs = require("fs");
const R = require("ramda");

const { getApplicasterConfig } = require("../../settings");
const { runInShellAsync } = require("../../shell");

async function runAndroid({ emulator, apkPath, zappAndroidPath, sleep }) {
  const env = R.compose(
    R.fromPairs,
    R.map(R.split("=")),
    R.split("\n")
  )(fs.readFileSync(`${zappAndroidPath}/.env`).toString());

  const { bundle_identifier: bundleIdentifier } = env;

  await runInShellAsync(
    `\${ANDROID_SDK}/emulator/emulator -avd ${emulator} -netdelay none -netspeed full -wipe-data & sleep ${sleep} \\
      && adb install ${apkPath} \\
      && adb reverse tcp:8081 tcp:8081 \\
      && adb shell am start -n ${bundleIdentifier}/com.applicaster.quickbrick.MainActivity \\
      && adb logcat *:S ReactNative:V ReactNativeJS:V`
  );
}

const runPlatform = {
  async ios({ zappApplePath }) {
    await runInShellAsync(
      `node_modules/.bin/react-native run-ios --project-path ${zappApplePath}/ZappiOS --scheme ZappiOS --configuration Debug`
    );
  },
  async tvos({ zappApplePath, tvOSEmulator }) {
    await runInShellAsync(
      `node_modules/.bin/react-native run-ios --project-path ${zappApplePath}/ZappTvOS --scheme ZappTvOS --configuration Debug --simulator "${tvOSEmulator}"`
    );
  },
  async android({ zappAndroidPath, androidMobileEmulator }) {
    await runAndroid({
      zappAndroidPath,
      emulator: androidMobileEmulator,
      sleep: 30,
      apkPath: `${zappAndroidPath}/app/build/outputs/apk/mobileGoogleQuickBrick/debug/app-mobile-google-quickbrick-debug.apk`,
    });
  },
  async androidTv({ zappAndroidPath, androidTVEmulator }) {
    await runAndroid({
      zappAndroidPath,
      emulator: androidTVEmulator,
      sleep: 15,
      apkPath: `${zappAndroidPath}/app/build/outputs/apk/tvGoogleQuickBrick/debug/app-tv-google-quickbrick-debug.apk`,
    });
  },
  async amazon({ zappAndroidPath, androidTVEmulator }) {
    await runAndroid({
      zappAndroidPath,
      emulator: androidTVEmulator,
      sleep: 15,
      apkPath: `${zappAndroidPath}/app/build/outputs/apk/tvAmazonQuickBrick/debug/app-tv-amazon-quickbrick-debug.apk`,
    });
  },
};

async function runApp({ platform }) {
  try {
    await runPlatform[platform](getApplicasterConfig());
    return true;
  } catch (e) {
    throw e;
  }
}

module.exports = { runApp };
