// @applicaster/zapp-react-native-utils/styleKeysUtils/index.js
import { platformSelect } from "../../reactUtils";

export const styleKeys = {
  loading_spinner_color: platformSelect({
    samsung_tv: "loading_spinner_color",
    lg_tv: "loading_spinner_color",
    tvos: "loading_spinner_color",
    ios: "loading_spinner_color",
    android_tv: "loading_spinner_color",
    android: "loading_spinner_color",
    default: "loading_spinner_color",
  }),
  loading_error_label: platformSelect({
    samsung_tv: "loading_error_label",
    lg_tv: "loading_error_label",
    tvos: "loading_error_label",
    ios: "loading_error_label",
    android: "loading_error_label",
    android_tv: "loading_error_label",
    default: "loading_error_label",
  }),
  background_color: platformSelect({
    samsung_tv: "background_color",
    lg_tv: "background_color",
    tvos: "background_color",
    ios: "background_color",
    android: "background_color",
    android_tv: "background_color",
    default: "background_color",
  }),
  player_overlay_color: platformSelect({
    samsung_tv: "player_overlay_color",
    lg_tv: "player_overlay_color",
    tvos: "player_overlay_color",
    ios: "player_overlay_color",
    android: "player_overlay_color",
    android_tv: "player_overlay_color",
    default: "player_overlay_color",
  }),
  main_text: platformSelect({
    samsung_tv: "main_text",
    lg_tv: "main_text",
    tvos: "main_text",
    ios: "main_text",
    android: "main_text",
    android_tv: "main_text",
    default: "main_text",
  }),
  active: platformSelect({
    samsung_tv: "active",
    lg_tv: "active",
    tvos: "active",
    ios: "active",
    android: "active",
    android_tv: "active",
    default: "active",
  }),
  focused_text_color: platformSelect({
    samsung_tv: "focused_text_color",
    lg_tv: "focused_text_color",
    tvos: "focused_text_color",
    ios: "focused_text_color",
    android: "focused_text_color",
    android_tv: "focused_text_color",
    default: "focused_text_color",
  }),
  background: platformSelect({
    samsung_tv: "background",
    lg_tv: "background",
    tvos: "background",
    ios: "background",
    android: "background",
    android_tv: "background",
    default: "background",
  }),
  header_text_color: platformSelect({
    samsung_tv: "header_text_color",
    lg_tv: "header_text_color",
    tvos: "header_text_color",
    ios: "header_text_color",
    android: "header_text_color",
    android_tv: "header_text_color",
    default: "header_text_color",
  }),
};
