// @applicaster/zapplicaster-cli/src/plugins/plugins_list.js
const camelize = require("camelize");

const plugin = ({ name, identifier, packageName, type = ["general"] }) => ({
  name,
  identifier,
  packageName,
  moduleName: camelize(identifier),
  type,
  version: "latest",
  configuration: null,
  default: true,
});

const defaultPlayer = plugin({
  name: "Quick Brick Default Player",
  identifier: "QuickBrickPlayerPlugin",
  type: ["player"],
  packageName: "@applicaster/zapp-react-native-default-player",
});

const defaultAppState = plugin({
  name: "Default App State Plugin",
  identifier: "quick-brick-default-app-state",
  packageName: "@applicaster/quick-brick-default-app-state",
});

const sideMenu = plugin({
  name: "Side Menu Plugin",
  identifier: "side_menu",
  type: ["menu"],
  packageName: "@applicaster/quick-brick-mobile-side-menu",
});

const topMenuBarTV = plugin({
  name: "Top Menu Bar TV",
  identifier: "top_menu_bar_tv",
  type: ["menu"],
  packageName: "@applicaster/quick-brick-top-menu-bar-tv",
});

const navBar = plugin({
  name: "Applicaster toolbar",
  type: ["nav_bar"],
  identifier: "applicaster_toolbar",
  packageName: "@applicaster/quick-brick-mobile-toolbar",
});

const defaultCellRenderer = plugin({
  name: "Default Cell Renderer",
  identifier: "default_cell_renderer",
  type: ["cell_builder"],
  packageName: "@applicaster/quick-brick-default-cell-renderer",
});

const applicaster2VodHook = plugin({
  name: "Applicaster2 Vod Hook",
  identifier: "applicaster2_vod_loader_hook",
  packageName: "@applicaster/quick-brick-applicaster2-vod-hook",
  type: ["general"],
});

const TvTransportControls = plugin({
  name: "TV Transport Controls",
  identifier: "tv_transport_controls",
  packageName: "@applicaster/quick-brick-tv-transport-controls",
  type: ["player_controls"],
});

const defaultTheme = plugin({
  name: "Base Theme",
  identifier: "default_theme",
  packageName: "@applicaster/quick-brick-base-theme",
  type: ["general"],
});

const defaultPlugins = {
  /* added for all platforms */
  all_platforms: [defaultAppState, defaultCellRenderer, applicaster2VodHook],
  /* device targets */
  mobile: [],
  tablet: [],
  universal: [],
  tv: [],
  /* platforms */
  ios: [],
  android: [],
  tvos: [],
  samsung_tv: [],
  lg_tv: [],
};

const fallbackPlugins = {
  player: {
    all_platforms: [defaultPlayer],
  },
  menu: {
    mobile: [sideMenu],
    universal: [sideMenu],
    tablet: [sideMenu],
    tv: [topMenuBarTV],
  },
  nav_bar: {
    mobile: [navBar],
    universal: [navBar],
    tablet: [navBar],
  },
  transport_controls: {
    samsung_tv: [TvTransportControls],
    lg_tv: [TvTransportControls],
  },
};

const mandatoryPlugins = [defaultTheme];

module.exports = {
  defaultPlugins,
  fallbackPlugins,
  mandatoryPlugins,
};
