// @applicaster/zapp-react-native-ui-components/Components/AudioPlayer/AudioPlayerLayout.js
// @flow
import React from "react";
import { View } from "react-native";

import { platformSelect } from "@applicaster/zapp-react-native-utils/reactUtils";

import { Artwork } from "./Artwork";
import { directionStyles } from "./helpers";

type Props = {
  artwork: ?string,
  config: {
    titleColor: "string",
    summaryColor: "string",
    backgroundColor: "string",
    backgroundImage: "string",
    isRTL: boolean,
  },
  children: React.ComponentType<any>,
};

export function AudioPlayerLayout({ artwork, config, children }: Props) {
  const { isRTL, backgroundColor } = config;

  const mainContainerStyles = platformSelect({
    tvos: {
      width: 1920,
      height: 1080,
      alignItems: "center",
      justifyContent: "center",
      flexDirection: directionStyles(isRTL).flexDirection,
      backgroundColor: backgroundColor,
    },
    android_tv: {
      width: 1920,
      height: 1080,
      alignItems: "center",
      justifyContent: "center",
      flexDirection: directionStyles(isRTL).flexDirection,
      backgroundColor: backgroundColor,
    },
    native: {},
    web: {
      position: "absolute",
      margin: "auto",
      display: "flex",
      flexWrap: "wrap",
      width: "100vw",
      flex: 1,
      alignItems: "center",
      justifyContent: "center",
      flexDirection: directionStyles(isRTL).flexDirection,
      backgroundColor: backgroundColor,
    },
  });

  const textContainerStyles = platformSelect({
    tvos: {
      width: 600,
      height: 362,
      marginHorizontal: 24,
      alignItems: directionStyles(isRTL).justifyContent,
    },
    android_tv: {
      width: 600,
      height: 362,
      marginHorizontal: 24,
      alignItems: directionStyles(isRTL).justifyContent,
    },
    web: {
      margin: 10,
      height: "100vh",
      alignItems: directionStyles(isRTL).justifyContent,
      justifyContent: "center",
    },
  });

  return (
    <View style={mainContainerStyles}>
      {!!artwork && <Artwork srcImage={artwork} config={config} />}
      <View style={textContainerStyles}>{children}</View>
    </View>
  );
}
