// @applicaster/zapp-react-native-ui-components/Components/AudioPlayer/AudioPlayer.js
// @flow
import React from "react";

import { imageSrcFromMediaItem } from "../MasterCell/MappingFunctions";

import { AudioPlayerLayout } from "./AudioPlayerLayout";
import { Channel } from "./Channel";
import { Title } from "./Title";
import { Summary } from "./Summary";
import { Runtime } from "./Runtime";
import { getPropertyFromEntryOrConfig } from "./helpers";

type Props = {
  audio_item: {
    title: string,
    summary: string,
    content: {
      src: string,
    },
    media_group: [
      {
        type: string,
        media_item: [
          {
            type: string,
            src: string,
            key: string,
          }
        ],
      }
    ],
    extensions: {
      audio_player_artwork_aspect_ratio: ?string,
      audio_player_background_image: ?string,
      audio_player_background_color: ?string,
      audio_player_channel_icon: ?string,
      audio_player_title_color: ?string,
      audio_player_summary_color: ?string,
      audio_player_rtl: ?boolean,
      start_time: ?string,
      end_time: ?string,
    },
  },
  plugin_configuration: {
    audio_player_background_color: ?string,
    audio_player_title_color: ?string,
    audio_player_summary_color: ?string,
    audio_player_rtl: ?string,
    audio_player_background_image: ?string,
    audio_player_artwork_aspect_ratio: ?string,
  },
  styles: {},
};

export function AudioPlayer({
  audio_item,
  plugin_configuration,
  styles,
}: Props) {
  const { extensions, title, summary } = audio_item;

  const getProp = getPropertyFromEntryOrConfig({
    entry: audio_item,
    plugin_configuration,
  });

  // Checking if we are recieving items from the DSP
  const titleColor = getProp("audio_player_title_color");
  const summaryColor = getProp("audio_player_summary_color");
  const backgroundColor = getProp("audio_player_background_color");
  const backgroundImage = getProp("audio_player_background_image");
  const artworkAspectRatio = getProp("audio_player_artwork_aspect_ratio");
  const channelIcon = getProp("audio_player_channel_icon");
  const rtlFlag = getProp("audio_player_rtl");
  const isRTL = rtlFlag === "1" || rtlFlag === "true" || rtlFlag === true;

  // we are recieving strings "0" or "1" from the matifest

  const config = {
    titleColor,
    summaryColor,
    backgroundColor,
    backgroundImage,
    isRTL,
  };

  const artwork = imageSrcFromMediaItem(audio_item, [artworkAspectRatio]);

  return (
    <AudioPlayerLayout
      artwork={artwork}
      aspectRatio={artworkAspectRatio}
      config={config}
    >
      <Channel srcImage={channelIcon} config={config} />
      <Title title={title} config={config} />
      <Summary summary={summary} config={config} />
      <Runtime {...extensions} config={config} />
    </AudioPlayerLayout>
  );
}
