// @applicaster/zapp-react-dom-app/App/Loader/loadLocale.js
/* global tizen */
import { platformSelect } from "@applicaster/zapp-react-native-utils/reactUtils";
import * as R from "ramda";
import { QuickBrickCommunicationModule } from "../../Polyfills/QuickBrickCommunicationModule";

type Locale = {
  language: string,
  country: string,
};

/**
 * Supported platforms: [Samsung] TODO: LG
 * This function is adding languageLocale and countryLocale to the QuickBrickCommunicationModule
 * @return Promise
 * Warning! This function mutates QuickBrickCommunicationModule
 */
export function loadLocale() {
  const getDefault = () => Promise.resolve();
  const getSamsungLocale = () =>
    new Promise((resolve, reject) => {
      if (typeof tizen === "undefined") resolve();
      try {
        tizen.systeminfo.getPropertyValue(
          "LOCALE",
          (locale: Locale) => {
            const { language } = locale;
            const [languageLocale, countryLocale] = R.compose(
              R.map(R.toLower),
              R.split("_")
            )(language);

            QuickBrickCommunicationModule.languageLocale = languageLocale;
            QuickBrickCommunicationModule.countryLocale = countryLocale;
            resolve();
          },
          err => {
            reject(new Error(err));
          }
        );
      } catch (err) {
        reject(new Error(err));
      }
    });

  return platformSelect({
    samsung_tv: getSamsungLocale,
    lg_tv: getDefault,
    default: getDefault,
  })();
}
