// @applicaster/zapp-react-native-tvos-cli-template/index.js
const START_SCRIPT = "./node_modules/.bin/react-native start";

const BUILD_IOS_SCRIPT =
  "mkdir -p ios && node_modules/.bin/react-native bundle --platform ios --dev false --entry-file index.js --bundle-output ./ios/main.jsbundle --reset-cache"; // eslint-disable-line max-len

const { resolve } = require("path");

const workspaceDependencies = [
  { name: "@applicaster/zapp-react-native-tvos-app" },
  { name: "@applicaster/zapp-react-native-tvos-ui-components" },
  { name: "@applicaster/zapp-react-native-ui-components" },
  { name: "applicaster/babel-preset-quick-brick", type: "devDependencies" },
];

const quickBrickDependencies = [{ name: "@applicaster/quick-brick-core" }];

const packageJson = require("./package.json");

const mapDependencies = ({ name, type = "dependencies" }) => ({
  name,
  version: packageJson.dependencies[name],
  type,
});

module.exports = {
  files: {
    templatePath: resolve(__dirname, "./template"),
    filesToCopy: [".babelrc", "_package.json"],
    filesToRender: [
      {
        templatePath: resolve(__dirname, "./template/index.js.ejs"),
        filePath: "index.js",
      },
      {
        templatePath: resolve(__dirname, "./template/config/index.js.ejs"),
        filePath: config => `config/index.${config.platform}.js`,
      },
    ],
  },
  dependencies: [
    ...workspaceDependencies.map(mapDependencies),
    ...quickBrickDependencies.map(mapDependencies),
    { name: "buffer", version: "^5.1.0", type: "dependencies" },
    { name: "events", version: "^3.0.0", type: "dependencies" },
    { name: "react", version: "16.8.3", type: "dependencies" },
    { name: "react-native", version: "0.59.10", type: "dependencies" },
    { name: "stream", version: "^0.0.2", type: "dependencies" },
    { name: "timers", version: "^0.1.1", type: "dependencies" },
    {
      name: "remote-redux-devtools",
      version: "^0.5.12",
      type: "devDependencies",
    },
  ],
  scripts: [
    {
      name: "start",
      command: START_SCRIPT,
    },
    {
      name: "build",
      command: BUILD_IOS_SCRIPT,
    },
  ],
};
