// @applicaster/zapp-react-native-android-tv-ui-components/Components/Header/Header.js
// @flow
import * as React from "react";
import * as R from "ramda";
import { Text } from "react-native";
import { ContainerWrapperView } from "../ContainerWrapperView";
import { fixColorHexCode } from "@applicaster/zapp-react-native-utils/stylesUtils";

type Props = {
  styles: any,
  cellStyles: any,
  title: string,
  visible: boolean,
  remoteConfigurations: {},
};

const baseStyles = {
  marginHorizontal: 20,
  marginBottom: 10,
  flex: 1,
};

const textStyles = (textColor, fontSize, fontFamily) => ({
  color: textColor || "white",
  fontSize: fontSize || 36,
  fontFamily: fontFamily || "Helvetica, Arial, Sans-serif",
});

export function HeaderComponent({ styles, title, visible }: Props) {
  if (!visible) {
    return null;
  }

  const textColor = fixColorHexCode(
    R.path(["android_tv", "header_font_style", "color"], styles)
  );

  const fontSize = Number(
    R.path(["android_tv", "header_font_style", "font_size"], styles)
  );

  const fontFamily = R.path(
    ["android_tv", "header_font_style", "font"],
    styles
  );

  const headerTitle = () => (
    <Text style={textStyles(textColor, fontSize, fontFamily)}>{title}</Text>
  );

  return (
    <ContainerWrapperView style={R.merge(baseStyles, styles)}>
      {headerTitle}
    </ContainerWrapperView>
  );
}
