// @applicaster/zapp-react-native-ui-components/Components/Transitioner/config.android.js
// @flow
import { Dimensions, Easing } from "react-native";

function pushTransition(animatedValue, width, height) {
  return {
    duration: 220,
    easing: Easing.linear,
    from: {
      style: {
        position: "absolute",
        width,
        height,
        opacity: animatedValue.interpolate({
          inputRange: [0.0, 1.0],
          outputRange: [1.0, 0.0],
        }),
      },
    },
    to: {
      style: {
        position: "absolute",
        width,
        height,
        opacity: animatedValue.interpolate({
          inputRange: [0.0, 1.0],
          outputRange: [0.0, 1.0],
        }),
        transform: [
          {
            translateY: animatedValue.interpolate({
              inputRange: [0.0, 1.0],
              outputRange: [height, 0.0],
            }),
          },
        ],
      },
    },
  };
}

function backTransition(animatedValue, width, height) {
  return {
    duration: 220,
    easing: Easing.cubic,
    from: {
      style: {
        position: "absolute",
        width,
        height,
        opacity: animatedValue.interpolate({
          inputRange: [0.0, 1.0],
          outputRange: [1.0, 0.0],
        }),
        transform: [
          {
            translateY: animatedValue.interpolate({
              inputRange: [0.0, 1.0],
              outputRange: [0.0, height],
            }),
          },
        ],
      },
    },
    to: {
      style: {
        position: "absolute",
        width,
        height,
        opacity: animatedValue.interpolate({
          inputRange: [0.0, 1.0],
          outputRange: [0.0, 1.0],
        }),
      },
    },
  };
}

export function transitionConfig(animatedValue) {
  const { width, height } = Dimensions.get("window");

  return {
    push: pushTransition(animatedValue, width, height),
    back: backTransition(animatedValue, width, height),
  };
}
