// @applicaster/zapp-react-native-ui-components/Components/Transitioner/config.ios.js
// @flow
import { Dimensions, Easing } from "react-native";

function pushTransition(animatedValue, width, height) {
  return {
    duration: 500,
    easing: Easing.out(Easing.poly(5)),
    from: {
      style: {
        position: "absolute",
        width,
        height,
        opacity: animatedValue.interpolate({
          inputRange: [0.0, 0.9, 1.0],
          outputRange: [1.0, 1.0, 0.0],
        }),
        transform: [
          {
            translateX: animatedValue.interpolate({
              inputRange: [0.0, 1.0],
              outputRange: [0.0, -(width / 2)],
            }),
          },
        ],
      },
    },
    to: {
      style: {
        position: "absolute",
        width,
        height,
        transform: [
          {
            translateX: animatedValue.interpolate({
              inputRange: [0.0, 1.0],
              outputRange: [width, 0.0],
            }),
          },
        ],
      },
    },
  };
}

function backTransition(animatedValue, width, height) {
  return {
    duration: 500,
    easing: Easing.out(Easing.poly(5)),
    from: {
      style: {
        position: "absolute",
        width,
        height,
        transform: [
          {
            translateX: animatedValue.interpolate({
              inputRange: [0.0, 1.0],
              outputRange: [0.0, width],
            }),
          },
        ],
      },
    },
    to: {
      style: {
        position: "absolute",
        width,
        height,
        transform: [
          {
            translateX: animatedValue.interpolate({
              inputRange: [0.0, 1.0],
              outputRange: [-(width / 2), 0.0],
            }),
          },
        ],
      },
    },
  };
}

export function transitionConfig(animatedValue) {
  const { width, height } = Dimensions.get("window");

  return {
    push: pushTransition(animatedValue, width, height),
    back: backTransition(animatedValue, width, height),
  };
}
