// @applicaster/zapp-react-native-tvos-ui-components/Components/Header/Header.js
// @flow
import * as React from "react";
import { CellRendererResolver } from "@applicaster/zapp-react-native-ui-components/Components/CellRendererResolver";

type Props = {
  configuration: {
    style: string,
    data: {
      target: ?string,
    },
    cell_plugin_configuration_id: ?string,
  },
  item: any,
  plugins: [{ module: any, type: string, name: string }],
  styles: any,
  cellStyles: any,
  remoteConfigurations: {},
  isScreenWrappedInContainer: boolean,
};

export function HeaderComponent({
  configuration,
  item,
  plugins,
  styles,
  cellStyles,
  isScreenWrappedInContainer,
}: Props) {
  const { cell_plugin_configuration_id } = configuration;

  // Since CellRendererResolver expect a structure of a rivers' component,
  // we are contructing an object that looks exacly like it.
  // The "header" is supported in the built-in Default Cell Renderer plugin,
  const component = {
    component_type: "header",
    styles: {
      cell_plugin_configuration_id,
    },
  };

  const containerStyles = isScreenWrappedInContainer ? { marginLeft: 4 } : {};

  const Renderer = CellRendererResolver({
    component,
    plugins,
    styles: { ...styles, ...containerStyles },
    cellStyles,
  });

  return <Renderer item={item} />;
}
