// @applicaster/zapp-react-native-utils/appUtils/keyCodes/keys.ios.js
/* global tizen */
const isTizen = typeof tizen !== "undefined";

export const KEYS = {
  Enter: { code: "Enter", keyCode: 13 },
  Space: { code: "Space", keyCode: 32 },
  m: { code: "m", keyCode: 77 },
  ArrowLeft: { code: "ArrowLeft", keyCode: 37 },
  ArrowUp: { code: "ArrowUp", keyCode: 38 },
  ArrowRight: { code: "ArrowRight", keyCode: 39 },
  ArrowDown: { code: "ArrowDown", keyCode: 40 },
  Back: isTizen
    ? { code: "Back", keyCode: 10009 }
    : { code: "Backspace", keyCode: 8 },
  Play: isTizen
    ? { code: "MediaPlay", keyCode: 10252 }
    : { code: "PlayPause", keyCode: 179 },
  Pause: isTizen
    ? { code: "MediaPause", keyCode: 19 }
    : { code: "Pause", keyCode: 179 },
  Forward: isTizen
    ? { code: "MediaFastForward", keyCode: 417 }
    : { code: "Forward", keyCode: 176 },
  Rewind: isTizen
    ? { code: "MediaRewind", keyCode: 412 }
    : { code: "Rewind", keyCode: 177 },
  VolumeUp: isTizen
    ? { code: "VolumeUp", keyCode: 447 }
    : { code: "VolumeUp", keyCode: 175 },
  VolumeDown: isTizen
    ? { code: "VolumeDown", keyCode: 448 }
    : { code: "VolumeDown", keyCode: 174 },
  ToggleMute: isTizen
    ? { code: "VolumeMute", keyCode: 449 }
    : { code: "VolumeMute", keyCode: 173 },
  Stop: isTizen
    ? { code: "MediaStop", keyCode: 413 }
    : { code: "Q", keyCode: 81 },
};

export const ARROW_KEYS = [
  KEYS.ArrowUp,
  KEYS.ArrowLeft,
  KEYS.ArrowDown,
  KEYS.ArrowRight,
];
