// @applicaster/zapp-react-native-utils/appUtils/keyCodes/keys.android.js
export const KEYS = {
  Enter: { code: "Enter", keyCode: 13 },
  Back: { code: "Back", keyCode: 8 },
  Space: { code: "Space", keyCode: 32 },
  m: { code: "m", keyCode: 77 },
  ArrowLeft: { code: "ArrowLeft", keyCode: 37 },
  ArrowUp: { code: "ArrowUp", keyCode: 38 },
  ArrowRight: { code: "ArrowRight", keyCode: 39 },
  ArrowDown: { code: "ArrowDown", keyCode: 40 },
};

export const ARROW_KEYS = [
  KEYS.ArrowUp,
  KEYS.ArrowLeft,
  KEYS.ArrowDown,
  KEYS.ArrowRight,
];

export function keyEventMapper(keyCode) {
  switch (keyCode) {
    case 19: // ArrowUp
      return { code: "ArrowUp", keyCode: 38 };
    case 20: // ArrowDown
      return { code: "ArrowDown", keyCode: 40 };
    case 21: // ArrowLeft
      return { code: "ArrowLeft", keyCode: 37 };
    case 22: // ArrowRight
      return { code: "ArrowRight", keyCode: 39 };
    case 23: // Center
      return { code: "Enter", keyCode: 13 };
    case 4: // Back
      return { code: "Back", keyCode: 8 };
    default:
      return { code: "None", keyCode: 0 };
  }
}
