// @applicaster/zapp-react-native-utils/appUtils/keyCodes/keysByPlatform.js
const lg_tv = {
  Space: {},
  m: {},
  Back: { keyCode: 461 },
  Play: { keyCode: 415 },
  Pause: { code: "Pause", keyCode: 19 },
  Stop: { keyCode: 413 },
  Forward: { keyCode: 417 },
  Rewind: { keyCode: 412 },
  TogglePlayPause: {},
  VolumeUp: {},
  VolumeDown: {},
  ChannelUp: { keyCode: 33 },
  ChannelDown: { keyCode: 34 },
  ToggleMute: {},
  Exit: { code: "Exit", keyCode: 10182 },
  Red: { keyCode: 403 },
  Green: { keyCode: 404 },
  Yellow: { keyCode: 405 },
  Blue: { keyCode: 406 },
  Sap: { keyCode: 1028 },
  Rec: { keyCode: 416 },
  QView: { keyCode: 1007 },
  CC: { keyCode: 460 },
  Fav: { keyCode: 1011 },
  Digit0: { code: "Digit0", keyCode: 48 },
  Digit1: { code: "Digit1", keyCode: 49 },
  Digit2: { code: "Digit2", keyCode: 50 },
  Digit3: { code: "Digit3", keyCode: 51 },
  Digit4: { code: "Digit4", keyCode: 52 },
  Digit5: { code: "Digit5", keyCode: 53 },
  Digit6: { code: "Digit6", keyCode: 54 },
  Digit7: { code: "Digit7", keyCode: 55 },
  Digit8: { code: "Digit8", keyCode: 56 },
  Digit9: { code: "Digit9", keyCode: 57 },
};

const samsung_tv = {
  m: { code: "m", keyCode: 77 },
  Back: { code: "Back", keyCode: 10009 },
  Play: { code: "MediaPlay", keyCode: 415 },
  TogglePlayPause: { code: "MediaPlayPause", keyCode: 10252 },
  Pause: { code: "MediaPause", keyCode: 19 },
  Forward: { code: "MediaFastForward", keyCode: 417 },
  Rewind: { code: "MediaRewind", keyCode: 412 },
  VolumeUp: { code: "VolumeUp", keyCode: 447 },
  VolumeDown: { code: "VolumeDown", keyCode: 448 },
  ToggleMute: { code: "VolumeMute", keyCode: 449 },
  Stop: { code: "MediaStop", keyCode: 413 },
  Exit: { code: "Exit", keyCode: 10182 },
};

const web = {
  Enter: { code: "Enter", keyCode: 13 },
  Space: { code: "Space", keyCode: 32 },
  m: { code: "m", keyCode: 77 },
  ArrowLeft: { code: "ArrowLeft", keyCode: 37 },
  ArrowUp: { code: "ArrowUp", keyCode: 38 },
  ArrowRight: { code: "ArrowRight", keyCode: 39 },
  ArrowDown: { code: "ArrowDown", keyCode: 40 },
  Back: { code: "Backspace", keyCode: 8 },
  Play: { code: "PlayPause", keyCode: 179 },
  TogglePlayPause: { code: "PlayPause", keyCode: 179 },
  Pause: { code: "Pause", keyCode: 179 },
  Forward: { code: "Forward", keyCode: 176 },
  Rewind: { code: "Rewind", keyCode: 177 },
  VolumeUp: { code: "VolumeUp", keyCode: 175 },
  VolumeDown: { code: "VolumeDown", keyCode: 174 },
  ToggleMute: { code: "VolumeMute", keyCode: 173 },
  Stop: { code: "Q", keyCode: 81 },
  Exit: { code: "E", keyCode: 8 },
};

export { lg_tv, samsung_tv, web };
