// @applicaster/zapplicaster-cli/src/templates/newWorkspaceTemplate/index.js
/* eslint-disable max-len */

const { resolve } = require("path");

const template = {
  files: {
    templatePath: resolve(__dirname, "./files"),
    filesToCopy: [".gitignore", "_package.json", ".eslintrc", "README.md"],
    filesToRender: [],
  },
  dependencies: [
    { name: "@babel/core", version: "^7.2.2", type: "devDependencies" },
    { name: "@babel/preset-env", version: "^7.3.1", type: "devDependencies" },
    { name: "@babel/preset-flow", version: "^7.0.0", type: "devDependencies" },
    { name: "@babel/preset-react", version: "^7.0.0", type: "devDependencies" },
    { name: "babel-core", version: "^7.0.0-bridge.0", type: "devDependencies" },
    { name: "babel-eslint", version: "^8.0.3", type: "devDependencies" },
    { name: "babel-jest", version: "^23.6.0", type: "devDependencies" },
    { name: "babel-preset-env", version: "^1.7.0", type: "devDependencies" },
    {
      name: "babel-preset-react-native-stage-0",
      version: "^1.0.1",
      type: "devDependencies",
    },
    { name: "ejs", version: "^2.5.7", type: "devDependencies" },
    { name: "enzyme", version: "3.3.0", type: "devDependencies" },
    {
      name: "enzyme-adapter-react-16",
      version: "1.1.1",
      type: "devDependencies",
    },
    { name: "enzyme-adapter-utils", version: "1.3.0", type: "devDependencies" },
    { name: "enzyme-to-json", version: "3.3.0", type: "devDependencies" },
    { name: "eslint", version: "^4.13.1", type: "devDependencies" },
    {
      name: "eslint-config-standard",
      version: "^11.0.0",
      type: "devDependencies",
    },
    { name: "eslint-plugin-babel", version: "^4.1.2", type: "devDependencies" },
    {
      name: "eslint-plugin-flowtype",
      version: "^2.47.1",
      type: "devDependencies",
    },
    {
      name: "eslint-plugin-import",
      version: "^2.12.0",
      type: "devDependencies",
    },
    {
      name: "eslint-plugin-jest",
      version: "^21.15.1",
      type: "devDependencies",
    },
    { name: "eslint-plugin-node", version: "^6.0.1", type: "devDependencies" },
    {
      name: "eslint-plugin-promise",
      version: "^3.7.0",
      type: "devDependencies",
    },
    { name: "eslint-plugin-react", version: "^7.8.2", type: "devDependencies" },
    {
      name: "eslint-plugin-react-native",
      version: "^3.2.0",
      type: "devDependencies",
    },
    {
      name: "eslint-plugin-standard",
      version: "^3.1.0",
      type: "devDependencies",
    },
    { name: "flow-bin", version: "^0.107.0", type: "devDependencies" },
    { name: "jest", version: "^23.6.0", type: "devDependencies" },
    { name: "jest-extended", version: "^0.6.0", type: "devDependencies" },
    { name: "jest-html-reporter", version: "^2.3.0", type: "devDependencies" },
    { name: "lerna", version: "^2.11.0", type: "devDependencies" },
    {
      name: "metro-react-native-babel-preset",
      version: "^0.51.1",
      type: "devDependencies",
    },
    { name: "ramda", version: "^0.25.0", type: "devDependencies" },
    {
      name: "react-addons-test-utils",
      version: "^15.6.2",
      type: "devDependencies",
    },
    { name: "react-dom", version: "^16.4.0", type: "devDependencies" },
    {
      name: "react-router-enzyme-context",
      version: "^1.2.0",
      type: "devDependencies",
    },
    { name: "react-test-renderer", version: "16.4.1", type: "devDependencies" },
    {
      name: "regenerator-runtime",
      version: "^0.13.1",
      type: "devDependencies",
    },
    { name: "react", version: "16.0.0", type: "dependencies" },
    { name: "react-native", version: "0.50.4", type: "dependencies" },
  ],
  scripts: [
    { name: "test", command: "node_modules/.bin/jest -w=2" },
    {
      name: "test:coverage",
      command: "node_modules/.bin/jest -w=2 --coverage",
    },
    {
      name: "lint",
      command: "node_modules/.bin/eslint --ignore-path .gitignore --ext .js .",
    },
    {
      name: "zapplicaster",
      command: "node ./packages/zapplicaster-cli/index.js",
    },
    {
      name: "clean:node",
      command:
        "rm -rf node_modules yarn.lock development-app/node_modules development-app/yarn.lock packages/**/node_modules packages/**/yarn.lock && yarn",
    },
    {
      name: "start",
      command:
        "node_modules/.bin/react-native start --config ./../../../../rn-cli.config.js",
    },
    {
      name: "run:ios",
      command:
        "node_modules/.bin/react-native run-ios --project-path $ZAPP_APP_PATH --scheme Zapp-App --configuration Debug",
    },
    { name: "start:ios", command: "yarn run:ios & yarn start" },
    {
      name: "run:tvos",
      command:
        'node_modules/.bin/react-native run-ios --project-path $ZAPP_APP_PATH --scheme ZappTvOS --configuration Debug --simulator "Apple TV"', // eslint-disable-line quotes
    },
    { name: "start:tvos", command: "yarn run:tvos & yarn start" },
    {
      name: "run:android",
      command:
        "node_modules/.bin/react-native run-android --project-path $ZAPP_APP_PATH --configuration Debug",
    },
    { name: "start:android", command: "yarn run:android & yarn start" },
    {
      name: "run:android_tv",
      command:
        "node_modules/.bin/react-native run-android_tv --project-path $ZAPP_APP_PATH --scheme QuickBrickAndroidTV --configuration Debug",
    },
    { name: "start:android_tv", command: "yarn run:android_tv & yarn start" },
    {
      name: "build:ios",
      command:
        " mkdir -p ios && node_modules/.bin/react-native bundle --config ./../../../../rn-cli.config.js --platform ios --dev false --entry-file development-app/index.js --bundle-output ios/main.jsbundle",
    },
    {
      name: "build:tvos",
      command:
        "mkdir -p tvos && node_modules/.bin/react-native bundle --config ./../../../../rn-cli.config.js --platform ios --dev false --entry-file development-app/index.js --bundle-output ios/main.jsbundle",
    },
    {
      name: "build:android",
      command:
        "mkdir -p android && node_modules/.bin/react-native bundle --config ./../../../../rn-cli.config.js --platform android --dev false --entry-file development-app/index.js --bundle-output android/main.jsbundle",
    },
    {
      name: "build:android_tv",
      command:
        "mkdir -p android_tv && node_modules/.bin/react-native bundle --config ./../../../../rn-cli.config.js --platform android --dev false --entry-file development-app/index.js --bundle-output android/main.jsbundle",
    },
  ],
};

module.exports = template;
