// @applicaster/zapplicaster-cli/src/commands/reloadConfig/index.js
const { taskRunner } = require("../../taskRunner");

const {
  prerequisitesChecker,
} = require("../prepareWorkspace/prerequisitesChecker");
const { configurator } = require("../prepareWorkspace/configurator");
const { workspaceCreator } = require("../prepareWorkspace/workspaceCreator");
const {
  configFilesRetriever,
} = require("../prepareWorkspace/configFilesRetriever");
const {
  runtimeConfigurationUrlsCreator,
} = require("../prepareWorkspace/runtimeConfigurationUrlsCreator");

const reloadConfigTask = {
  name: "reload configuration",
  startMessage: "We're reloading the configuration for your app",
  prerequisitesChecker,
  configurator,
  steps: [
    {
      start: "Creating workspace",
      run: workspaceCreator,
      error: "Could not create workspace",
      completion: "Workspace created !",
    },
    {
      start: "Retrieving configuration files",
      run: configFilesRetriever,
      error: "Could not retrieve configuration files",
      completion: "Configuration files retrieved !",
    },
    {
      start: "Creating runtime URLs configuration file",
      run: runtimeConfigurationUrlsCreator,
      error: "Could not create runtime_configuration_urls.json",
      completion: "Runtime configuration URLs file created",
    },
  ],
};

module.exports = { reloadConfig: taskRunner(reloadConfigTask) };
