// @applicaster/zapp-react-native-tvos-ui-components/Components/DefaultNavigation/NavigationTitle/index.js
// @flow
import * as React from "react";
import { View, Text } from "react-native";
import { fixColorHexCode } from "@applicaster/zapp-react-native-utils/stylesUtils";

type Props = {
  title: string,
  summary: string,
  menuProps: {
    styles: {
      summary_color: string,
      summary_font_size: string,
      summary_font: string,
      title_color: string,
      title_font: string,
      title_font_size: string,
    },
  },
};
const navigationTitleStyle = {
  flex: 1,
  justifyContent: "center",
  alignItems: "center",
};
export function NavigationTitle({
  title = "",
  summary = "",
  menuProps,
}: Props) {
  const {
    styles: {
      summary_color,
      summary_font_size,
      summary_font,
      title_color,
      title_font,
      title_font_size,
    },
  } = menuProps;

  const titleStyles = {
    fontSize: parseInt(title_font_size, 10),
    color: fixColorHexCode(title_color) || "rgba(0, 0, 0, 0)",
    fontFamily: title_font,
  };
  const summaryStyles = {
    fontSize: parseInt(summary_font_size, 10),
    color: fixColorHexCode(summary_color) || "rgba(0, 0, 0, 0)",
    fontFamily: summary_font,
  };
  return (
    <View style={navigationTitleStyle}>
      <Text style={titleStyles}>{title}</Text>
      <Text numberOfLines={2} ellipsizeMode={"tail"} style={summaryStyles}>
        {summary}
      </Text>
    </View>
  );
}
