// @applicaster/zapp-react-dom-ui-components/Components/Collection/Collection.js
// @flow
import * as React from "react";
import { View } from "react-native";
import { Cell } from "../Cell";

import { Header } from "../Header";
import { ContainerWrapperView } from "../ContainerWrapperView";

type Props = {
  items: [],
  componentId: string,
  component: {},
  CellRenderer: React.componentType<{ item: Object, state: string }>,
  header: {
    title: string,
    data: { target: string | null },
    styles: ?{},
    visible: boolean,
  },
  styles: {},
  shouldScrollVertically: ?([any]) => boolean,
  shouldScrollHorizontally: ?([any]) => boolean,
  navigator: {},
  isScreenWrappedInContainer: ?boolean,
};

export function CollectionComponent({
  items,
  componentId,
  component,
  CellRenderer,
  header,
  styles,
  shouldScrollVertically,
  shouldScrollHorizontally,
  navigator,
  isScreenWrappedInContainer,
}: Props) {
  return (
    <View>
      <Header {...header} />
      <ContainerWrapperView style={styles}>
        {() =>
          items.map((item, index) => (
            <Cell
              componentId={componentId}
              component={component}
              CellRenderer={CellRenderer}
              item={item}
              key={index}
              index={index}
              hasHeader={header.visible}
              shouldScrollHorizontally={shouldScrollHorizontally}
              shouldScrollVertically={shouldScrollVertically}
              hasPreferredFocus={index === 0}
              navigator={navigator}
              isScreenWrappedInContainer={isScreenWrappedInContainer}
            />
          ))
        }
      </ContainerWrapperView>
    </View>
  );
}
