// @applicaster/zapp-react-native-utils/navigationUtils/__tests__/fixtures.js
const dataSourceUrl =
  "applicaster://fetchData?type=APPLICASTER_ATOM_FEED&url=foo";

const media_group = [
  {
    type: "image",
    media_item: [
      {
        src: "http://image.png",
        key: "image_base",
        type: "image",
      },
    ],
  },
];

export const river = {
  id: "A1234",
  name: "home",
  type: "general_content",
  data: { source: null, type: null, connected: null, target: "B4567" },
  home: true,
  position: 1,
  styles: {},
  ui_components: [],
  navigations: [],
};

export const riverPlugin = {
  id: "B0987",
  name: "ScreenPlugin",
  type: "screen_plugin",
  data: {},
  home: false,
  navigations: [],
};

export const riverFeed = {
  type: { value: "feed" },
  title: "nested river feed",
  riverId: "B4567",
  id: "1234",
  content: { src: dataSourceUrl },
  link: {
    type: "atom",
    rel: "self",
    href: dataSourceUrl,
  },
  media_group,
};

export const menuItem = {
  id: "A1234",
  assets: { icon: "https://icon.png" },
  data: { source: null, type: null, connected: null, target: "B4567" },
  general: { home: false, new: false },
  position: 1,
  styles: {},
  title: "River",
  type: "label",
};

export const navBarItem = {
  id: "A1234",
  assets: { icon: "https://icon.png" },
  data: { source: null, type: null, connected: null, target: "B4567" },
  general: { home: false, new: false },
  position: 1,
  styles: {},
  title: "Settings",
  type: "button",
};

export const article = {
  author: { name: "Author Name" },
  content: { type: "html" },
  id: "19560",
  link: {
    rel: "alternate",
    type: "text/html",
    href: "appscheme://presentRN?plugin=Article",
  },
  media_group,
  summary: "",
  title: "This is the title of the article",
  type: { value: "link" },
};

export const articleWithoutHTMLContent = {
  ...article,
  content: { type: "foo" },
  type: { value: "article" },
};

export const content = {
  id: "1234",
};

export const vod = {
  id: "video1234",
  type: { value: "video" },
};

export const channel = {
  id: "channel1234",
  type: { value: "channel" },
};

export const playable = {
  id: "playable1234",
  content: { type: "video/hls" },
};

export const audio = {
  id: "audio_1234",
  type: { value: "audio" },
};

export const urlScheme = {
  url: "appscheme://present?screen_id=A1234",
};
