// @applicaster/zapp-react-native-android-tv-ui-components/Components/Cell/CellStyles/CellStyle2/index.js
import { cellStyle2 as cellStyles } from "../../styles";

const viewTree = state => [
  {
    type: "View",
    style: cellStyles.container,
    elements: [
      {
        type: "View",
        style:
          state === "focused"
            ? cellStyles.focusedContainer
            : cellStyles.selectedContainer,
        elements: [
          {
            type: "View",
            style: cellStyles.imageContainer,
            elements: [
              {
                type: "Image",
                style: cellStyles.image,
                data: [
                  {
                    func: "image_src_from_media_item",
                    args: ["thumbnail-small"],
                    propName: "uri",
                  },
                ],
              },
              {
                type: "View",
                style: cellStyles.titleContainer,
                elements: [
                  {
                    type: "Text",
                    style: cellStyles.title,
                    data: [
                      {
                        func: "path",
                        args: ["title"],
                        propName: "label",
                      },
                    ],
                    additionalProps: {
                      numberOfLines: 2,
                    },
                  },
                ],
              },
              {
                type: "View",
                style: cellStyles.subTitleContainer,
                elements: [
                  {
                    type: "Text",
                    style: cellStyles.subTitle,
                    data: [
                      {
                        func: "path",
                        args: ["summary"],
                        propName: "label",
                      },
                    ],
                    additionalProps: {
                      numberOfLines: 1,
                    },
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
];

export const cellStyle2 = {
  content_types: {
    default: {
      states: {
        default: viewTree("default"),
        focused: viewTree("focused"),
      },
    },
  },
};
