// @applicaster/zapplicaster-cli/src/commands/publishPlugin/index.js
const { taskRunner } = require("../../taskRunner");

const {
  prerequisitesChecker,
} = require("../prepareWorkspace/prerequisitesChecker");

const { configurator } = require("./configurator");
const { npmPublish } = require("./npmPublish");
const { generateManifest } = require("./generateManifest");
const { zappifestPublish } = require("./zappifestPublish");
const { commitToGit } = require("./commitToGit");

const publishPluginTask = {
  name: "publish plugin",
  startMessage: "Publishing plugin",
  prerequisitesChecker,
  configurator,
  steps: [
    {
      start: "publishing to npm",
      run: npmPublish,
      error: "Couldn't publish to npm",
      completion: "Your plugin was successfully published to NPM",
    },
    {
      start: "generating manifests",
      run: generateManifest,
      error: "Couldn't generate the manifests",
      completion:
        "Your manifests have been generated in the manifests folder of your plugin",
    },
    {
      start: "publishing manifest",
      run: zappifestPublish,
      error: "Couldn't push zappifest to Zapp",
      completion: "Zappifest uploaded ! Your plugin is available in Zapp",
    },
    {
      start: "Committing to git",
      run: commitToGit,
      error: "Couldn't commit changes to git",
      completion: "Done !",
    },
  ],
};

module.exports = { publishPlugin: taskRunner(publishPluginTask) };
