// @applicaster/zapp-react-native-tvos-app/Context/LayoutContext/index.js
import * as R from "ramda";
import { createContext } from "@applicaster/zapp-react-native-utils/reactUtils/createContext";

// This boollean give information that current screen should not be allow item to be focusded
const SCREEN_FOCUS_BLOCKED = "screenFocusBlocked";

// Pass infirmation that screen was created and constructor was called
const SCREEN_DID_LOAD = "screenDidLoad";

// Pass infomation that screen was presented after initial screen animation
const SCREEN_DID_APPEAR = "screenDidAppear";

// Pass infomation that screen did disappear
const SCREEN_DID_DISAPPEAR = "screenDidDisappear";

// Pass information that screen focused on some item inside it
const SCREEN_FOCUSED_ON_ITEM = "screenFocusedOnItem";

function noop() {}

const initialContext = {
  screenFocusBlocked: false,
  screenDidLoad: noop,
  screenDidAppear: noop,
  screenFocusedOnItem: noop,
  screenDidDisappear: noop,
};
function stateValidator({ property, value }) {
  if (property === SCREEN_FOCUS_BLOCKED) {
    return R.is(Boolean, value);
  }
  if (property === SCREEN_DID_APPEAR) {
    return R.is(Function, value);
  }
  if (property === SCREEN_DID_LOAD) {
    return R.is(Function, value);
  }
  if (property === SCREEN_FOCUSED_ON_ITEM) {
    return R.is(Function, value);
  }
  if (property === SCREEN_DID_DISAPPEAR) {
    return R.is(Function, value);
  }
  return false;
}

export const LayoutContext = createContext(initialContext, stateValidator);
