// @applicaster/zapp-react-dom-ui-components/Components/River/RiverContainer/RiverContainer.js
import * as React from "react";
import { View } from "react-native";

import { VariableOffsetContainer } from "../../VariableOffsetContainer";
import { ContainerWrapperView } from "../../ContainerWrapperView";

type Props = {
  setRiverOffsetUpdater: any => void,
  children: React.ComponentType<any>,
  isScreenWrappedInContainer: ?boolean,
  marginTopOffset: Number,
};

const containerStyle = { minHeight: "calc(100vh - 120px)" };

const riverStyles = (marginTopOffset = 0) => ({
  marginTop: 235 + marginTopOffset,
  zIndex: 0,
  transformStyle: "preserve-3d",
});

const overflowContainer = {
  overflow: "hidden",
};

export function RiverContainerComponent({
  setRiverOffsetUpdater,
  isScreenWrappedInContainer,
  marginTopOffset,
  children,
}: Props) {
  return (
    <View style={overflowContainer}>
      <ContainerWrapperView
        style={riverStyles(marginTopOffset)}
        isScreenWrappedInContainer={isScreenWrappedInContainer}
      >
        {() => (
          <VariableOffsetContainer
            baseStyles={containerStyle}
            saveUpdaterFunction={setRiverOffsetUpdater}
            transitionDelay={400}
            easing="cubic-bezier(0, 1, 0.5, 1)"
          >
            {styles => <div style={styles}>{children}</div>}
          </VariableOffsetContainer>
        )}
      </ContainerWrapperView>
    </View>
  );
}
