// @applicaster/zapp-react-native-android-tv-ui-components/Components/Carousel/index.js
// @flow
import * as React from "react";
import * as R from "ramda";
import { StyleSheet, View } from "react-native";
import SnapCarousel from "react-native-snap-carousel";

import { useNavigation } from "@applicaster/zapp-react-native-utils/reactHooks/navigation";
import { Focusable } from "@applicaster/zapp-react-native-ui-components/Components/Focusable";
import { Cell } from "../Cell";

import { useInitialFocus } from "@applicaster/zapp-react-native-utils/focusManager";

type Props = {
  data: [any],
  focused: boolean,
  keyExtractor: (any) => string,
  parentFocus: {},
  nextFocusUp: React.Ref,
  nextFocusDown: React.Ref,
  component: {},
  header: { title: string, visible: boolean },
  CellRenderer: React.Node,
};

const carouselStyles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "row",
  },
});

export function Carousel(props: Props) {
  const {
    data,
    focused,
    nextFocusUp,
    nextFocusDown,
    component,
    header,
    CellRenderer,
  } = props;

  const carouselRef = React.useRef(null);
  const [activeIndex, setActiveIndex] = React.useState(0);
  const carouselFocusable = React.useRef(null);
  const navigator = useNavigation();

  useInitialFocus(focused, carouselFocusable);

  const moveFocus = (direction) => {
    if (direction === "right") {
      carouselRef.current.snapToNext(true);
    } else {
      carouselRef.current.snapToPrev(true);
    }
  };

  const navigateTo = (item) => {
    const targetScreen = component?.data?.target;
    const targetRoute = R.merge({ screen_type: targetScreen }, item);
    navigator.push(targetRoute);
  };

  const onFocus = (element, { direction, reset } = {}) => {
    if (direction && !reset) {
      moveFocus(direction);
    }
  };

  // eslint-disable-next-line react/prop-types
  const renderItem = ({ item, index }) => {
    const cellProps = {
      parentFocus: props.parentFocus,
      focused,
      componentId: component?.id,
      item,
      CellRenderer,
      navigator,
      hasHeader: header.visible,
    };

    return (
      <View>
        <Cell {...cellProps} />
      </View>
    );
  };

  const onSnapToItem = (index) => {
    setActiveIndex(index);
  };

  return (
    <View style={carouselStyles.container}>
      <Focusable
        onFocus={onFocus}
        onPress={() => navigateTo(data?.[activeIndex])}
        id="carousel-focusable"
        ref={carouselFocusable}
        nextFocusRight={carouselFocusable}
        nextFocusLeft={carouselFocusable}
        nextFocusUp={nextFocusUp}
        nextFocusDown={nextFocusDown}
      >
        {() => null}
      </Focusable>

      <SnapCarousel
        onSnapToItem={onSnapToItem}
        ref={carouselRef}
        data={data}
        horizontal
        renderItem={renderItem}
        itemWidth={842}
        sliderWidth={1920}
        loop
        inactiveSlideScale={1}
        inactiveSlideOpacity={1}
        scrollEnabled={false}
      />
    </View>
  );
}
